File: modules/my-rust-module/ios/MyRustModule.swift
public class MyRustModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MyRustModule")

    Constants([
      "PI": Double.pi
    ])

    Events("onChange")

    Function("hello") {
      return "Hello world! 👋"
    }

    AsyncFunction("setValueAsync") { (value: String) in
      self.sendEvent("onChange", [
        "value": value
      ])
    }

    // Adding the readQr function
    AsyncFunction("readQr") { (path: String) -> String in
      let cPath = path.cString(using: .utf8)!  // Convert Swift String to C-String
      let resultPtr = read_qr(cPath)  // Assuming read_qr is exposed and callable
      let result = String(cString: resultPtr)  // Convert back to Swift String from C-String
      free(resultPtr)  // Don't forget to free the memory if it's allocated in Rust
      return result
    }

    View(MyRustModuleView.self) {
      Prop("name") { (view: MyRustModuleView, prop: String) in
        print(prop)
      }
    }
  }
}

File: modules/my-rust-module/android/src/main/java/expo/modules/myrustmodule/MyRustModule.kt
package expo.modules.myrustmodule

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class MyRustModule : Module() {
  companion object {
    // Load the native library
    init {
      System.loadLibrary("native_rust_lib")  // Ensure this matches your actual library name
    }
  }

  // Declare external functions matching the Rust implementations
  external fun readQr(path: String): Long
  external fun freeQr(qrPtr: Long)
  external fun getQrContent(qrPtr: Long): String
  external fun getQrBounds(qrPtr: Long): IntArray

  override fun definition() = ModuleDefinition {
    Name("MyRustModule")

    Constants("PI" to Math.PI)

    Events("onChange")

    Function("hello") {
      "Hello world! 👋"
    }

    AsyncFunction("readQr") { path: String ->
      val qrPtr = readQr(path)
      val content = getQrContent(qrPtr)
      val bounds = getQrBounds(qrPtr)
      freeQr(qrPtr)
      mapOf("content" to content, "bounds" to bounds)
    }

    AsyncFunction("setValueAsync") { value: String ->
      sendEvent("onChange", mapOf("value" to value))
    }

    View(MyRustModuleView::class) {
      Prop("name") { view: MyRustModuleView, prop: String ->
        println(prop)
      }
    }
  }
}
